_posts/2018-10-04-js-testing.md

```diff
@@ -0,0 +1,206 @@
+
+![Coverage]({{"/assets/test_coverage.png" | absolute_url}})
+
+In this article we will talk about the testing tools we use for our javascript projects.
+
+* [Jest](https://jestjs.io/)
+Delightful JavaScript Testing, Jest is used by Facebook to test all JavaScript code including  
+React applications. One of Jest's philosophies is to provide an integrated "zero-configuration" experience. 
+
+* [Enzyme](https://airbnb.io/enzyme/)
+Is a JavaScript Testing utility for React that makes it easier to assert, manipulate, and traverse your  
+React Components' output.
+
+Ok, those are cool names but what can we do with them?  
+Here are some examples on how to use these tools:
+
+We have a utility function that validates dates are in the following format DD/MM/YYYY
+
+``` javascript
+function areInvalidDates(values){
+    let status = [];
+    if (values.length > 0) {
+      status = values.map(value => {
+        if (value.split("/").length !== 3 ||
+            value.split("/")[2].length !== 4 ||
+            value.split("/")[1].length !== 2 ||
+            value.split("/")[0].length !== 2
+            ) {
+            return 'invalid';
+        } else {
+          return 'valid';
+        }
+      });
+    }
+    return status.includes('invalid');
+  }
+
+  // the test
+  describe('utils', function() {
+    it('should have a function that validates an array of date string values', function () {
+      const valuesWithInvalid = ["01/02/1990", "01/04/22", "02/02/2000"];
+      const hasOneInvalidYear = utils.areInvalidDates(valuesWithInvalid);
+      expect(hasOneInvalidYear).toBe.true;
+
+      const valuesWithAnotherInvalid = ["01/02/1990", "01/4/2000", "02/02/2000"];
+      const hasInvalidDay = utils.areInvalidDates(valuesWithAnotherInvalid);
+      expect(hasInvalidDay).toBe.true;
+
+      const valuesWithSomeInvalid = ["1/02/1990", "01/04/2000", "02/02/2000"];
+      const hasInvalidMonth = utils.areInvalidDates(valuesWithSomeInvalid);
+      expect(hasInvalidMonth).toBe.true;
+
+      const valuesWithAllValid = ["01/02/1990", "01/04/2001", "02/02/2000"];
+      const allValid = utils.areInvalidDates(valuesWithAllValid);
+      expect(allValid).toBe.false;
+    });
+  });
+```
+
+This first example makes use of Jest library to set what we expect to see with each set of values.
+The behavior of the function on each case is evaluated and we can set the expected result according  
+the params.
+
+Now how does Jest and Enzyme work with React components? :thinking:
+Let's say we have a login box component with fields for email and password:
+
+```javascript
+
+class LoginBox extends Component {
+
+  constructor() {
+    super();
+    this.state = {
+      email: '',
+      password: ''
+    };
+  }
+
+  saveInput(name, event) {
+    if ((event.keyCode || event.which) === 13) {
+      this.submitForm(event);
+      return;
+    }
+    let state = {};
+    state[name] = event.target.value;
+    this.setState(state);
```

Comment:
Does this work?

```suggestion
    this.setState({ name: event.target.value });
```

Reply:
the current code works, but we can go for your example 🤔